File: backend/app/schemas/tarea.py
from datetime import date, datetime
from pydantic import BaseModel, Field, model_validator
from typing import Literal, Optional


ENTIDAD_FKS = ["cliente_id", "empresa_id", "cotizacion_id", "nota_venta_id", "factura_id", "producto_id"]


class TareaIn(BaseModel):
    titulo: str = Field(min_length=1, max_length=255)
    descripcion: str | None = None
    due_date: date
    asignado_id: int
    cliente_id: int | None = None
    empresa_id: int | None = None
    cotizacion_id: int | None = None
    nota_venta_id: int | None = None
    factura_id: int | None = None
    producto_id: int | None = None

    @model_validator(mode="after")
    def max_una_fk(self):
        fks = [getattr(self, f) for f in ENTIDAD_FKS]
        if sum(1 for v in fks if v is not None) > 1:
            raise ValueError("Solo se puede vincular a UNA entidad")
        return self


class TareaPatch(BaseModel):
    titulo: str | None = Field(default=None, min_length=1, max_length=255)
    descripcion: str | None = None
    due_date: date | None = None
    asignado_id: int | None = None


class TareaOut(BaseModel):
    id: int
    titulo: str
    descripcion: str | None
    due_date: date
    estado: Literal["pendiente", "hecha", "descartada"]
    motivo_descarte: str | None
    origen: Literal["manual", "auto"]
    tipo_regla: str | None
    prioridad_derivada: Literal["vencida", "hoy", "futura"]
    asignado_id: int
    asignado_nombre: str
    creado_por_id: int | None
    cliente_id: int | None
    empresa_id: int | None
    cotizacion_id: int | None
    nota_venta_id: int | None
    factura_id: int | None
    producto_id: int | None
    completada_at: datetime | None
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class DescartarIn(BaseModel):
    motivo: str = Field(min_length=1, max_length=255)


class ReasignarIn(BaseModel):
    asignado_id: int


class MisPendientesOut(BaseModel):
    vencidas: int
    hoy: int
    futuras: int
    total: int
    tareas: list[TareaOut]

File: frontend/src/types/tarea.ts
export type TareaEstado = 'pendiente' | 'hecha' | 'descartada';
export type TareaOrigen = 'manual' | 'auto';
export type PrioridadDerivada = 'vencida' | 'hoy' | 'futura';
export type AsignadoRol = 'vendedor' | 'admin' | 'owner';
export type EntidadTipo = 'cliente' | 'empresa' | 'cotizacion' | 'nota_venta' | 'factura' | 'producto';

export interface Tarea {
  id: number;
  titulo: string;
  descripcion: string | null;
  due_date: string;
  estado: TareaEstado;
  motivo_descarte: string | null;
  origen: TareaOrigen;
  tipo_regla: string | null;
  prioridad_derivada: PrioridadDerivada;
  asignado_id: number;
  asignado_nombre: string;
  creado_por_id: number | null;
  cliente_id: number | null;
  empresa_id: number | null;
  cotizacion_id: number | null;
  nota_venta_id: number | null;
  factura_id: number | null;
  producto_id: number | null;
  completada_at: string | null;
  created_at: string;
  updated_at: string;
}

export interface TareaPage {
  items: Tarea[];
  total: number;
  page: number;
  page_size: number;
}

export interface MisPendientes {
  vencidas: number;
  hoy: number;
  futuras: number;
  total: number;
  tareas: Tarea[];
}

export interface ReglaTarea {
  id: number;
  tipo: string;
  activa: boolean;
  offset_dias: number;
  asignado_rol: AsignadoRol;
}

export interface TareaCreateInput {
  titulo: string;
  descripcion?: string;
  due_date: string;
  asignado_id: number;
  cliente_id?: number;
  empresa_id?: number;
  cotizacion_id?: number;
  nota_venta_id?: number;
  factura_id?: number;
  producto_id?: number;
}

export interface TareaFiltros {
  asignado_id?: number;
  estado?: TareaEstado;
  prioridad_derivada?: PrioridadDerivada;
  cliente_id?: number;
  empresa_id?: number;
  cotizacion_id?: number;
  nota_venta_id?: number;
  factura_id?: number;
  producto_id?: number;
  origen?: TareaOrigen;
  page?: number;
  page_size?: number;
}
